sunpy/io/special/genx.py

```diff
@@ -0,0 +1,140 @@
+            if dim == 2 and arr_size[0] == 1:
+                # For when structures has been defined with 2 dim but only has one:
+                # bb = replicate({tata: 1, bebe:2, casa:'asa'}, 3)
+                # dd = replicate({tata: 1, bebe:2, casa:'asa'}, 1, 3)
+                # print,size(bb), size(dd)
+                #            1           3           8           3
+                #            2           1           3           8           3
+                dim = 1
+                arr_size[0] = arr_size[1]
+            if arr_size[-1] > 1:
+                tagdict[tt] = np.array([read_struct_skeleton(xdrdata)] * arr_size[-1]).reshape(arr_size[0:-2])
+            else:
+                tagdict[tt] = read_struct_skeleton(xdrdata)
+        else:
+            tagdict[tt] = [dim] + arr_size
+    return tagdict
+
+def unsuported_dtype(dtype):
+    raise ValueError("Can't read {} data type".format(dtype))
+
+def struct_to_data(xdrdata, subskeleton):
+    """"
+    Converts the dictionary with the keys and IDL's size output to
+    the data stored in the xdrdata.
+    """
+    #http://www.harrisgeospatial.com/docs/SIZE.html
+    types_dict = {
+        2: (xdrdata.unpack_int, np.int16), # int
+        3: (xdrdata.unpack_int, np.int32), # long
+        4: (xdrdata.unpack_float, np.float32),
+        5: (xdrdata.unpack_double, np.float64),
+        6: (partial(unsuported_dtype, 'Complex'), np.float32), # FIXME: How to read complex?
+        7: (xdrdata.unpack_string, None),
+        9: (partial(unsuported_dtype, 'Double Complex'), np.float64), # FIXME: How to read Double complex?
+        12: (xdrdata.unpack_uint, np.uint16), # unsign int
+        13: (xdrdata.unpack_uint, np.uint32), # unsign Long int
+        14: (partial(unsuported_dtype, 'Long 64'), np.int64), # Long64          # FIXME: How to do 64?
+        15: (partial(unsuported_dtype, 'Unsign Long 64'), np.uint64), # unsign Long64   # FIXME: How to do 64?
+    }
+    for key in subskeleton:
+        if isinstance(subskeleton[key], OrderedDict):
+            struct_to_data(xdrdata, subskeleton[key])
+        elif isinstance(subskeleton[key], np.ndarray):
+            testlist = list()
+            struct_shape = subskeleton[key].shape
+            for elem in subskeleton[key].flatten():
+                elem2 = copy.deepcopy(elem)
+                struct_to_data(xdrdata, elem2)
+                testlist.append(elem2)
+            subskeleton[key] = np.array(testlist).reshape(struct_shape)
+        else:
+            sswsize = subskeleton[key]
+            sswtype = sswsize[-2]
+            if sswsize[0] == 0:
+                subskeleton[key] = types_dict[sswtype][0]()
+            else:
+                subskeleton[key] = np.array(xdrdata.unpack_farray(sswsize[-1], types_dict[sswtype][0]),
+                                            dtype=types_dict[sswtype][1]).reshape(sswsize[1:-2][::-1])
+
+def read_genx(filename):
+    """
+    solarsoft genx file reader
```

Comment:
This could be a little more descriptive, i.e. so someone who has never seen SSW knows what it is!